[1단계 - 로또 구현] 다니(이다은) 미션 제출합니다.

da-nyee opened the pull request:
안녕하세요, 코니! 다니입니다 🙌
이번에 처음 뵙는데, 코드 리뷰 잘 부탁드립니다! 😊

저는 로또 미션을 통해 모든 원시값과 문자열을 포장, enum 활용, 일급 콜렉션 사용을 경험했습니다.
특히, 모든 원시값과 문자열 포장은 처음 해봤는데 페어와 함께한 덕분에 금방 적용할 수 있었습니다 !

또한, 로또를 구현하며 TDD를 다시 이해할 수 있었는데요.
무조건 테스트 코드를 먼저 작성하고, 다음에 프로덕션 코드를 작성하는 연습을 했습니다.

제가 미션을 진행하며 궁금한 점을 몇 가지 정리했는데요, 질문 드려도 괜찮을까요??

## 궁금한 점 1.
메소드명을 변경할 때 커밋 메시지 타입을 `style`로 적었는데요, refactor와 style 사이에서 고민을 했습니다.
혹시 코니는 보통 어떤 타입을 사용하나요? 둘 중 어떤 게 더 나은 건가요??

## 궁금한 점 2.
로또 번호 자동 생성 기능은 `Collections.shuffle()`을 이용해서 무작위로 섞고 `subList()`를 사용해서 상위 6개 인덱스만 선택해서 발급되게 구현했는데요. 현재 코드에는 `new ArrayList<>()`의 인자로 `subList()`가 들어가서 매번 문제 없이 랜덤값이 반환되는데, `new ArrayList<>()`가 없이 `subList()`만 작성하면 모든 경우에 동일한 List가 반환됩니다.
```
예를 들어, 로또 번호를 3번 발급해야 하면

[1, 2, 3, 4, 5, 6]
[1, 2, 3, 4, 5, 6]
[1, 2, 3, 4, 5, 6]

이렇게 출력됩니다.
```
혹시 `subList()`가 객체 생성을 매번 따로 안하는 건가 싶어서 `hashCode()`로 hash 주소도 확인해봤는데, 매 순간마다 다른 값을 반환했습니다. 그래서 `subList()`도 객체를 따로 생성하는 거 같은데, 왜 `ArrayList<>()`로 복사를 하지 않으면 전부 같은 subList가 나오는 건지 이유를 알 수 있을까요??
```
  public static List<LottoNumber> shuffleLottoNumbers() {
      Collections.shuffle(lottoNumbers);
      List<LottoNumber> splitLottoNumbers = new ArrayList<>(lottoNumbers.subList(MINIMUM_RANGE, MAXIMUM_RANGE));
      return Collections.unmodifiableList(splitLottoNumbers);
  }
```

## 궁금한 점 3.
프로그래밍 요구사항에 따라 모든 원시값과 문자열을 포장하던 도중에 궁금증이 생겼습니다.
아래 코드처럼 Integer는 Wrapper 타입이라 따로 포장을 안했는데, 이런 경우에도 포장하는 게 필요할까요??
```
Integer count = lottoStatistics.get(rank);
```

제 질문은 여기까지 입니다!

코드 리뷰를 하면서 코드에 부족한 점이나 잘못된 부분이 있다면 부담 없이 전부 지적해주세요!
피드백 기다리고 있겠습니다. 귀한 시간 내주셔서 감사합니다!! 🥰

pobiconan committed:
refactor: remove spark library, html

tributetothemoon committed:
docs: 기능 구현 목록 작성

tributetothemoon committed:
feat: LottoNumber 생성자 구현

tributetothemoon committed:
feat: LottoNumber 타입, 범위 검증 로직 구현

tributetothemoon committed:
feat: LottoTicket 생성 로직 구현

tributetothemoon committed:
feat: Money 클래스 생성자 및 검증 로직 구현

tributetothemoon committed:
refactor: InvalidNumberRangeException 생성자 리팩토링

- 메세지를 인자로 받아 생성되도록 리팩토링하였습니다.

tributetothemoon committed:
feat: LottoNumberRepository 생성

tributetothemoon committed:
style: LottoTicket.of(List<String>) 메소드 명은 valueOf로 변경

tributetothemoon committed:
feat: 랜덤하게 로또 번호를 추첨하는 기능 구현

tributetothemoon committed:
feat: LottoTickets 생성

- LottoTicket에 대한 일급 컬렉션

tributetothemoon committed:
feat: 주어진 금액에 따라 로또를 구매하는 기능 구현

tributetothemoon committed:
feat: Rank 클래스 구현

- 맞춘 개수에 따른 상금에 대한 정보를 가지는 enum 클래스 Rank를 구현했습니다.

tributetothemoon committed:
feat: 당첨번호(WinningNumber) 생성 및 검증 로직

tributetothemoon committed:
feat: 당첨 통계를 구하는 로직 구현

tributetothemoon committed:
feat: 수익률 계산 기능 구현

tributetothemoon committed:
feat: 구입 금액, 당첨 번호, 보너스 번호 입력 기능 구현

tributetothemoon committed:
docs: 기능 구현 목록 추가

tributetothemoon committed:
fix: 여러 개의 자동 추첨 로또가 모두 같은 번호인 문제 수정

- Shallow Copy로 인해 발생하는 버그로, Deep Copy를 통해 버그를 해결하였습니다.

tributetothemoon committed:
feat: 구입 금액을 입력받고 로또를 사는 과정까지 Controller 로직 구현

tributetothemoon committed:
feat: 당첨 번호와 보너스 볼 입력 과정까지 Controller 로직 구현

tributetothemoon committed:
feat: 당첨통계에서 일치 개수 구하는 기능 구현

tributetothemoon committed:
feat: 당첨 통계에 따른 상금 계산 기능 추가

tributetothemoon committed:
refactor: Money의 인스턴스 변수 money 타입 int -> long으로 변경

tributetothemoon committed:
feat: 당첨 통계 일치 개수 출력 기능 구현

tributetothemoon committed:
feat: 당첨 통계 수익률 출력 기능 구현

tributetothemoon committed:
fix: 보너스 볼이 포함되는 경우 무조건 2등이 되는 문제 해결

tributetothemoon committed:
refactor: 에러 메세지 상수화

tributetothemoon committed:
style: Money 예외 메세지 상수 오타 수정

tributetothemoon committed:
refactor: InputView 에러 메세지 상수화

tributetothemoon committed:
refactor: 일부 에러 메세지 접두사 [ERROR] 누락 수정

da-nyee commented:
# 📚 다니의 학습 로그

## [TDD] 테스트 주도 개발 - 5
### 내용
- 테스트가 가능한 클래스와 불가능한 클래스를 구분 지음
- 테스트가 가능한 클래스에 대해 TDD를 수행함
  - 먼저, 실패하는 테스트 코드를 작성함
  - 다음으로, 컴파일 에러가 나지 않을 정도의 프로덕션 코드를 작성함
  - 다음으로, 기능 요구사항 대로 결과가 나타나는 프로덕션 코드를 작성함
  - 마지막으로, 리팩토링을 통해 코드 컨벤션을 확인하고 클린 코드가 되도록 변경함
### 링크
- https://github.com/da-nyee/java-lotto/tree/step1/src/test/java/domain

## [Java] 일급 컬렉션 - 4
### 내용
- LottoTicket, LottoTickets 객체를 도출하여 `규칙 8: 일급 콜렉션을 쓴다.`를 지킴
- 지금까지 잘못 알고 있던 개념을 다시 확인하고 이해함
### 링크
- [일급 컬렉션 (First Class Collection)의 소개와 써야할 이유](https://jojoldu.tistory.com/412)
- https://github.com/da-nyee/java-lotto/blob/step1/src/main/java/domain/LottoTicket.java
- https://github.com/da-nyee/java-lotto/blob/step1/src/main/java/domain/LottoTickets.java

## [Java] 원시값 및 문자열 포장 - 5
### 내용
- LottoNumber, Money, Profit 객체를 도출하여 `규칙 3: 모든 원시값과 문자열을 포장한다.`를 지킴
- 지난 미션까지는 방법을 잘 몰랐으나, 이번 미션에서 직접 코드로 구현하며 습득함
### 링크
- https://github.com/da-nyee/java-lotto/blob/step1/src/main/java/domain/LottoNumber.java
- https://github.com/da-nyee/java-lotto/blob/step1/src/main/java/domain/Money.java
- https://github.com/da-nyee/java-lotto/blob/step1/src/main/java/domain/Profit.java

## [Java] Enum - 4
### 내용
- Rank 객체를 도출하여 `java enum을 적용해 프로그래밍을 구현한다.`를 지킴
- 여태 enum은 상수를 모아두는 용도로만 사용했는데, 이번 기회를 통해 enum의 다양한 활용 방식을 이해함
### 링크
- [Java Enum 활용기](https://woowabros.github.io/tools/2017/07/10/java-enum-uses.html)
- https://github.com/da-nyee/java-lotto/blob/step1/src/main/java/domain/Rank.java

## [Java] 스트림 - 5
### 내용
- 가능한 외부 반복자를 지양하고, 내부 반복자를 지향하려고 노력함
- 스트링에 익숙해지고자 여러 스트림 API를 사용함
### 링크
- [[Java] 스트림 (Stream)](https://da-nyee.github.io/posts/java-stream/)

## [Java] 제네릭 - 2
### 내용
- 런타임 시 제네릭 정보가 지워진다는 것을 알게 됨
- 따라서, 메소드 오버로딩이 불가능함
### 링크
- [Name clash: Two methods have the same erasure](https://coderanch.com/t/706161/java/clash-methods-erasure)

## [Java] 초기화 블럭 - 3
### 내용
- 초기화 블럭과 static 초기화 블럭의 존재를 알게 됨
- 둘의 차이를 이해함
### 링크
- [[Java-14]초기화와 초기화블럭(Initialize Block)](https://kamang-it.tistory.com/entry/Java-14%EC%B4%88%EA%B8%B0%ED%99%94%EC%99%80-%EC%B4%88%EA%B8%B0%ED%99%94%EB%B8%94%EB%9F%ADInitialize-Block)

choihz commented on src/test/java/domain/LottoNumberTest.java:
```diff
@@ -0,0 +1,45 @@
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class LottoNumberTest {
+    @DisplayName("같은 숫자인 경우 정상 처리된다")
```
같은 숫자인 경우 무엇이 정상 처리된다는 것일까요? 이름만 보고도 테스트가 하는 역할을 명확히 알 수 있다면 특히나 테스트가 실패했을 때 빠르게 상황을 파악할 수 있답니다!

da-nyee replied:
[`a0dad49`](https://github.com/woowacourse/java-lotto/pull/248/commits/a0dad492ba15156c375f9933fb79c720fcec09f9) 수정했습니다!

choihz commented on src/test/java/domain/LottoNumberTest.java:
```diff
@@ -0,0 +1,45 @@
+        assertThat(first).isNotEqualTo(second);
+    }
+
+    @DisplayName("45를 초과한 수가 입력될 경 예외를 발생시킨다.")
```
```suggestion
    @DisplayName("45를 초과한 수가 입력될 경우 예외를 발생시킨다.")
```
😎

da-nyee replied:
[`e0b924e`](https://github.com/woowacourse/java-lotto/pull/248/commits/e0b924ee92d47e738a0b73c23bafee6edc6acc78) 수정했습니다!

choihz commented on src/main/java/domain/LottoPurchase.java:
```diff
@@ -0,0 +1,16 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class LottoPurchase {
```
`LottoPurchase`라는 이름만 보고는 어떤 역할을 하는 객체인지 단번에 이해하긴 어렵네요. 돈을 주면 로또를 돌려주니 `LottoMachine` 같은 이름은 어떨까요? 더 좋은 이름이 있을 수도 있구요.

da-nyee replied:
[`9bd543f`](https://github.com/woowacourse/java-lotto/pull/248/commits/9bd543fe09246d0715fb24a81c5441bae1fc0cf8) 수정했습니다!

제가 기존에 사용한 `LottoPurchase`보다 코니가 제시해준 `LottoMachine`이 훨씬 좋은 이름 같아요! 추천 감사합니다 😆

choihz commented on src/main/java/domain/LottoTicket.java:
```diff
@@ -0,0 +1,52 @@
+
+    public static LottoTicket of(List<LottoNumber> shuffleLottoNumbers) {
+        return new LottoTicket(shuffleLottoNumbers);
+    }
```
1. `List<LottoNumber>`를 매개변수로 받는 생성자도 있고 정적 팩터리 메서드도 있군요. 밖에서 `List<LottoNumber>`로 `LottoTicket`을 생성하고자 할 때 어떤 것을 사용할지는 어떻게 판단해야 할까요? 정적 팩터리 메서드가 굳이 필요할까요?
2. 매개변수로 들어오는 `LottoNumber`들이 섞였는지 안 섞였는지는 `LottoTicket`의 관심사가 아닐 것 같아요. 따라서 매개변수 이름에 `shuffle`이 들어갈 필요는 없어 보여요.

da-nyee replied:
1. [`b7d487c`](https://github.com/woowacourse/java-lotto/pull/248/commits/b7d487c13344a1073feb9669e7bc866634b014ac) 수정했습니다!
2. [`0557c76`](https://github.com/woowacourse/java-lotto/pull/248/commits/0557c760c8f4c68d2bc24b7c6fa273ffefe44614) 수정했습니다!

1번을 리팩토링 하면서 궁금한 점이 생겼습니다!

Q. 정적 팩토리 메소드는 언제 활용하면 좋은 방법인가요??

저는 이번에 페어 프로그래밍을 하면서 정적 팩토리 메소드를 처음 알게 되었는데요. 그래서 아직 이런 방법도 있구나 정도로만 이해한 수준입니다. 생성자처럼 단지 `new` 키워드만 붙여서 반환하는 경우에는 생성자를 쓰는 게 낫고, 메소드 내부에서 어떤 처리를 해야 하면 정적 팩토리 메소드를 사용하면 되는 건가요??

choihz replied:
https://johngrib.github.io/wiki/static-factory-method-pattern/

⟪이펙티브 자바⟫에서는 정적 팩터리 메서드의 장점을 다음과 같이 소개하고 있어요.

1. 이름이 있으므로 생성자에 비해 가독성이 좋다.
2. 호출할 때마다 새로운 객체를 생성할 필요가 없다.
3. 하위 자료형 객체를 반환할 수 있다.
4. 형인자 자료형(parameterized type) 객체를 만들 때 편하다.

저도 위와 같은 장점을 누릴 수 있는 경우에 정적 팩터리 메서드를 사용하는 것은 좋다고 생각해요! 그런데 대부분의 경우에는 `of()`, `valueOf()` 처럼 별로 의미 없는 이름 + 단순한 생성자 리턴의 조합으로 사용하고 계신 것 같더라고요 ㅎㅎ 저는 웬만하면 생성자를 사용하는 것이 더 명확하다고 생각해서 선호하지만, 이것 역시 취향 차이입니다. 다니도 조금 더 공부해 보시고 언제 정적 팩터리 메서드를 사용했을 때 코드가 더 좋아지는지 느끼면서 본인만의 생각을 쌓아 나가시면 돼요.

choihz commented on src/main/java/domain/LottoTicket.java:
```diff
@@ -0,0 +1,52 @@
+        if (lottoNumbers.size() != TOTAL_LOTTO_NUMBER_COUNT) {
+            throw new IllegalArgumentException(ERROR_INVALID_NUMBER_COUNT);
+        }
+    }
```
`is`나 `has` 등으로 시작하는 메서드는 보통 메서드의 내용이 단순히 조건식인 경우가 많아요. 특히 위의 정적 팩터리 메서드에서 필터링에 `isNoneDuplicationNumberOrThrow()`를 사용하고 있는데, 필터링 도중에 예외가 발생할 수도 있겠군요 😱 괜찮은 설계일까요?
또한, 중복 확인 로직이 다소 복잡한데요, 우리가 이미 알고 있는 `Set`을 활용해 보면 어떨까요?

da-nyee replied:
[`6382155`](https://github.com/woowacourse/java-lotto/pull/248/commits/6382155a73b57f14f54cf3e06e125682630280da) 수정했습니다!

이 내용도 리팩토링 하면서 궁금한 점이 생겼는데요.

Q. 필터링 도중에 예외를 발생시키는 건 나쁜 설계인가요?? 미리 검증을 하면 생성되지 않을 객체가 생성돼서 그런 건가요??

choihz replied:
음 저는 필터링을 시켰을 때 예외가 발생한다면 예상치 못한 동작이라고 생각할 거 같아요... 이것도 역시 취향 차이일 수 있으려나요 🤔

choihz commented on src/main/java/domain/LottoTickets.java:
```diff
@@ -0,0 +1,20 @@
+import java.util.Collections;
+import java.util.List;
+
+public class LottoTickets {
```
이 일급 컬렉션은 하는 일이 딱히 없네요 🤔

da-nyee replied:
[`5be0e88`](https://github.com/woowacourse/java-lotto/pull/248/commits/5be0e8861355d05ab3f5e626273d0be9af4619ae) 수정했습니다!

choihz commented on src/main/java/domain/WinningNumber.java:
```diff
@@ -0,0 +1,28 @@
+    private static final String ERROR_DUPLICATED_NUMBER = "[ERROR] 보너스 볼은 당첨 번호와 중복될 수 없습니다.";
+
+    private final LottoTicket winningNumbers;
+    private final LottoNumber bonusBall;
```
`LottoTicket winningNumbers` + `LottoNumber bonusBall` = `WinningNumber`?
객체의 이름이 역할을 잘 나타내고 있는지 고민해 볼까요?

da-nyee replied:
[`6caf39d`](https://github.com/woowacourse/java-lotto/pull/248/commits/6caf39d3c47bb7fb20b27105c455df64b077e07b) 수정했습니다!

choihz commented on src/main/java/domain/WinningNumber.java:
```diff
@@ -0,0 +1,28 @@
+    public WinningNumber(LottoTicket winningNumbers, LottoNumber bonusBall) {
+        this.winningNumbers = winningNumbers;
+        validateBonusBall(bonusBall);
+        this.bonusBall = bonusBall;
```
```suggestion
        validateBonusBall(bonusBall);
        this.winningNumbers = winningNumbers;
        this.bonusBall = bonusBall;
```
큰 의미는 없지만, 이 순서가 더 안정감 있게 느껴져요 🙄

da-nyee replied:
[`6caf39d`](https://github.com/woowacourse/java-lotto/pull/248/commits/6caf39d3c47bb7fb20b27105c455df64b077e07b) 수정했습니다!

choihz commented on src/main/java/domain/Statistics.java:
```diff
@@ -0,0 +1,39 @@
+
+    public Statistics(WinningNumber winningNumber, LottoTickets lottoTickets) {
+        for (Rank rank : Rank.values()) {
+            lottoStatistics.put(rank, ZERO);
```
매직 넘버를 상수로 분리하는 것의 가장 큰 장점은 단순한 숫자에 의미를 부여해 줄 수 있는 것이라고 생각해요. 그런데 `0`에 `ZERO`라는 이름을 붙이면 어떤 의미가 부여될까요? 왜 생성자에서 `Map`에 `0`을 넣는지 바로 이해가 가능하게 할 수 있지 않을까요?

da-nyee replied:
[`a0618ef`](https://github.com/woowacourse/java-lotto/pull/248/commits/a0618ef92a12a5b87476294ab5de0ac54afb8b6c) 수정했습니다!

choihz commented on src/main/java/domain/WinningNumber.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    public Rank calculateRank(LottoTicket lottoTicket) {
+        long correctNumber = lottoTicket.numbers().stream()
```
`correct`는 맞다/틀리다의 맞다를 의미해서, 여기선 적절하지 않아 보여요.

da-nyee replied:
[`71242f2`](https://github.com/woowacourse/java-lotto/pull/248/commits/71242f26aa70e2ccbc3fdbe40a30b750e3ae6077) 수정했습니다!

choihz commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.Arrays;
+
+public enum Rank {
+    FIRST(6, new Money("2000000000")),
```
```suggestion
    FIRST(6, new Money("2_000_000_000")),
```
가독성 향상! 실수 가능성 감소!

da-nyee replied:
[`15a7254`](https://github.com/woowacourse/java-lotto/pull/248/commits/15a725486e34991edeb0b02fbb0532c1f296133d) 수정했습니다!

choihz commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,45 @@
+    private static final long FIVE = 5;
+
+    long count;
+    Money reward;
```
```suggestion
    private final long count;
    private final Money reward;
```
`count`보다 더 좋은 이름이 있지 않을까요?

da-nyee replied:
[`fbbb2af`](https://github.com/woowacourse/java-lotto/pull/248/commits/fbbb2af55c7aefb636b20b2e03b3ef87e6fec500) 수정했습니다!

choihz commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,45 @@
+    public static Rank from(long count, boolean bonusBallMatch) {
+        if (bonusBallMatch && count == FIVE) {
+            return Rank.SECOND;
+        }
```
1. `bonusBallMatch`보다는 `calculateRank()`에서 사용한 `hasBonus`라는 이름이 훨씬 직관적이고 좋은 것 같은데요? `if hasBonus ...`가 되어서 아래 `if` 문도 훨씬 쉽게 읽히고요.
2. 위에서도 말씀 드린 것처럼, `FIVE`라는 이름은 전혀 의미를 더해 주지 못하고 있어요! `count == FIVE`가 대체 어떤 의미일까요?

da-nyee replied:
1. [`0f7080f`](https://github.com/woowacourse/java-lotto/pull/248/commits/0f7080f9af8784eedc734c094f9c44c777fef727) 수정했습니다!
2. [`ccecc83`](https://github.com/woowacourse/java-lotto/pull/248/commits/ccecc83216b6ebffcdbe9bb10265e2c37dfde0b7) 수정했습니다!

choihz commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,45 @@
+        return Arrays.stream(Rank.values())
+                .filter(item -> item.getCount() == count)
+                .findAny()
+                .orElse(NOTHING);
```
👍

choihz commented on src/test/java/domain/MoneyTest.java:
```diff
@@ -0,0 +1,29 @@
+    public void equalsTest() {
+        Money money1 = new Money("1000");
+        Money money2 = new Money("1000");
+        assertThat(money1).isEqualTo(money2);
```
```suggestion
        Money money1 = new Money("1000");
        Money money2 = new Money("1000");

        assertThat(money1).isEqualTo(money2);
```
이렇게 한 줄씩 띄워 주면 가독성이 훨씬 좋아져요.

da-nyee replied:
[`aaf8709`](https://github.com/woowacourse/java-lotto/pull/248/commits/aaf8709de3cc71767aa4fe1a77ff62ff285ec549) 수정했습니다!

choihz commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,43 @@
+    private final InputView inputView = InputView.getInstance();
+    private final OutputView outputView = OutputView.getInstance();
+
+    public void run() {
```
이 메서드 내에서도 조금씩 줄바꿈을 해 주시면 조금 더 숨을 쉬며 읽을 수 있을 것 같아요 😳

da-nyee replied:
[`31962df`](https://github.com/woowacourse/java-lotto/pull/248/commits/31962df9d1cc6ef2d48d194c6185e94a97b65d9f) 수정했습니다!

choihz commented on src/main/java/domain/LottoNumber.java:
```diff
@@ -0,0 +1,51 @@
+    private final int number;
+
+    public LottoNumber(String input) {
+        validateInputFormat(input);
```
`LottoNumber`라는 도메인 객체가 `input`을 받고 검증해 객체를 생성하고 있군요. 클래스 이름에도 숫자가 들어가는 만큼, 숫자만 받을 수 있게 해줘도 좋을 것 같아요. 물론 이건 저의 개인적인 취향일 수도 있어요.

da-nyee replied:
[`15a7254`](https://github.com/woowacourse/java-lotto/pull/248/commits/15a725486e34991edeb0b02fbb0532c1f296133d) 수정했습니다!

코니 말대로 클래스명에도 숫자가 들어가니까 문자열보다 숫자를 받는 게 통일성도 있고 맞는 거 같아요 !

choihz commented on src/main/java/domain/LottoNumberRepository.java:
```diff
@@ -0,0 +1,27 @@
+
+    public static List<LottoNumber> shuffleLottoNumbers() {
+        Collections.shuffle(lottoNumbers);
+        List<LottoNumber> splitLottoNumbers = new ArrayList<>(lottoNumbers.subList(MINIMUM_RANGE, MAXIMUM_RANGE));
```
https://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html#subList(int,%20int)

> Returns **a view of the portion of this list** between the specified fromIndex, inclusive, and toIndex, exclusive. **The returned list is backed by this list, so non-structural changes in the returned list are reflected in this list, and vice-versa.**

```java
 public List<E> subList(int fromIndex, int toIndex) {
    subListRangeCheck(fromIndex, toIndex, size);
    return new SubList(this, 0, fromIndex, toIndex);
}
```

`ArrayList`의 `subList()` 코드를 보면 위와 같이 `SubList`라는 타입을 리턴하고 있는데요,

```java
private class SubList extends AbstractList<E> implements RandomAccess {
    private final AbstractList<E> parent;
    private final int parentOffset;
    private final int offset;
    int size;

    SubList(AbstractList<E> parent, int offset, int fromIndex, int toIndex) {
        this.parent = parent;
        this.parentOffset = fromIndex;
        this.offset = offset + fromIndex;
        this.size = toIndex - fromIndex;
        this.modCount = ArrayList.this.modCount;
    }
...
}
```

놀랍게도 이 `SubList`는 내부적으로 부모 리스트를 참조하고 있네요. 따라서 `new ArrayList<>()`로 새롭게 만들어 주지 않으면, 계속 부모를 참조하면서 부모의 변경을 따라가게 돼요. `LottoNumberRepository`에서의 사용 맥락에서 보면 `List<LottoNumber> lottoNumbers`가 부모 리스트인데 이 리스트가 `Collections.shuffle()`에 의해 계속 섞이고, 그 변화를 모든 `subList`가 공유하게 되죠. 결국 계속 섞이다가 마지막에 섞인 결과가 `lottoNumbers`의 최종 상태가 될 텐데, 앞에서 생성된 모든 `subList`가 이 최종 상태의 `lottoNumbers`에서 0~6번 인덱스의 원소를 추출하게 됩니다. 이해가 되셨을까요? 😇

da-nyee replied:
아하 부모 리스트 참조 때문이었군요..!!! `subList()`를 쓸 때는 `new ArrayList<>()`로 새로 생성하는 게 좋겠네요! `LottoNumberRepository`로 구체적인 설명을 해주셔서 바로 이해가 됐습니다! 궁금증 해결해주셔서 감사합니다 🥰

choihz reviewed (changes requested):
안녕하세요 다니, 리뷰어 코니입니다 👋
전체적으로 깔끔하게 잘 구현해 주셨고, 책임이 적절하게 나뉘어서 이해하기 좋은 코드였어요. 원시값 포장, 일급 컬렉션 사용, 스트림 활용 등도 훌륭합니다 💯

질문에 답을 드리자면,
**1번**. 저는 지금은 커밋 메시지에 prefix를 작성하지 않고 있어요. 크게 중요한 건 아니라고 생각해서, 다니의 취향대로 하시면 될 것 같습니다.
**3번**. 요구사항에서의 원시값 포장과 `Integer` 타입이 Wrapper 타입이란 건 맥락이 조금 다른 것 같아요 ㅎㅎ 우선 여기선 굳이 필요하지 않아 보입니다.

2번 질문에 대한 답변은 해당 위치에 남겼어요~ 추가적인 질문이나 의견은 언제든지 코멘트 및 DM으로 알려주세요!

Commit:
test: LottoNumber 객체 비교 테스트 DisplayName, 메소드명 수정

Commit:
style: 오탈자 수정

Commit:
style: LottoPurchase에서 LottoMachine으로 클래스명 변경

- 객체의 역할을 더 잘 나타내는 이름으로 변경했습니다.

Commit:
style: ZERO에서 INITIAL_COUNT로 상수명 변경

Commit:
style: count에서 match로 변수명 변경

Commit:
style: correctNumber에서 match로 변수명 변경

Commit:
style: bonusBallMatch에서 hasBonus로 변수명 변경

Commit:
style: FIVE에서 SECOND_MATCH로 상수명 변경

Commit:
test: 가독성을 위해 줄 바꿈 추가

Commit:
refactor: 코드가 위에서 아래로 읽히도록 메소드 위치 변경

Commit:
refactor: 가독성을 위해 메소드 분리 및 줄 바꿈 추가

Commit:
refactor: validateBonusBall 메소드 호출 위치 변경

Commit:
style: shuffleLottoNumbers에서 lottoNumbers로 변수명 변경

Commit:
refactor: WinningNumber에서 WinningLotto로 클래스명 변경

- validateBonusBall 메소드의 매개변수에 winningNumbers도 추가했습니다.

Commit:
test: 메소드명 수정

Commit:
refactor: LottoTicket 생성 시 정적 팩토리 메소드 대신 생성자 사용

Commit:
refactor: 정적 팩토리 메소드 로직 수정

- isValidNumberCountOrThrow에서 validateNumberCount로 메소드명을
변경했습니다.
- isNoneDuplicationNumberOrThrow에서 validateDuplicationNumber로
메소드명을 변경했습니다.
- 로또 번호 중복 확인 로직에 filter 대신 Set을 사용했습니다.

Commit:
refactor: 메소드 위치 변경 및 하드코딩 상수화

Commit:
refactor: 숫자를 받는 객체들의 생성자 매개변수 타입을 변경

- LottoNumber의 경우 String에서 int로 변환했습니다.
- Money의 경우 String에서 long으로 변환했습니다.

Commit:
style: valueOf에서 of로 메소드명 변경

Commit:
refactor: 메소드 위치 변경

Commit:
refactor: LottoTickets 일급 컬렉션 삭제

- LottoTickets를 List<LottoTicket>으로 변경했습니다.

da-nyee commented:
안녕하세요, 코니 🙌
코드 리뷰랑 질문들에 대한 답변 감사합니다!

피드백 바탕으로 리팩토링 했는데, 피드백 보면서 궁금했던 내용은 해당 코멘트에 👀와 함께 궁금한 점을 적어두었습니다!
확인해주시면 감사하겠습니다 ~!

이번에도 코드 리뷰 잘 부탁드립니다 ✨

choihz commented on src/main/java/domain/LottoNumber.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    private void validateNumberRange(int number) {
+        if ((number < MIN_NUMBER_RANGE) || (number > MAX_NUMBER_RANGE)) {
```
```suggestion
        if (number < MIN_NUMBER_RANGE || number > MAX_NUMBER_RANGE) {
```
괄호는 없어도 될 것 같아요.

choihz commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,52 @@
+
+    private List<LottoTicket> createLottoTickets(Money budget) {
+        List<LottoTicket> lottoTickets = LottoMachine.buy(budget);
+        outputView.printLottoTicket(lottoTickets);
+        outputView.newLine();
```
이 정도는 `OutputView`에서 알아서 한다면 컨트롤러가 일일이 지시할 필요가 없지 않을까요?

choihz reviewed (approved):
안녕하세요, 다니!

피드백 반영 잘 해주셨어요 👍 1단계는 여기서 마칠게요. 2단계에서 만나요!
